### PR TITLE
Have tidy ensure that we document all `unsafe` blocks in libcore

### PR DESCRIPTION
**Pull request by oli-obk**
cc @rust-lang/libs 	

I documented a few and added ignore flags on the other files. We can incrementally document the files, but won't regress any files this way.

### CODE REVIEW

File: src/libcore/benches/ascii.rs [hunk 1]
```diff
@@ -142,6 +143,7 @@ benches! {
     }
 
     fn case08_fake_simd_u64(bytes: &mut [u8]) {
+        // SAFETY: transmuting a sequence of `u8` to `u32` is always fine
```
Comment (fbstj) [inline src/libcore/benches/ascii.rs]:
```suggestion
        // SAFETY: transmuting a sequence of `u8` to `u64` is always fine
```

File: src/libcore/num/f32.rs [hunk 2]
```diff
@@ -456,6 +457,7 @@ impl f32 {
     #[stable(feature = "float_bits_conv", since = "1.20.0")]
     #[inline]
     pub fn from_bits(v: u32) -> Self {
+        // SAFETY: `u32` is a plain old datatype so we can always transmute to it
```
Comment (fbstj) [inline src/libcore/num/f32.rs]:
```suggestion
        // SAFETY: `u32` is a plain old datatype so we can always transmute from it
```

File: src/libcore/num/f64.rs [hunk 3]
```diff
@@ -469,6 +470,7 @@ impl f64 {
     #[stable(feature = "float_bits_conv", since = "1.20.0")]
     #[inline]
     pub fn from_bits(v: u64) -> Self {
+        // SAFETY: `u64` is a plain old datatype so we can always transmute to it
```
Comment (fbstj) [inline src/libcore/num/f64.rs]:
```suggestion
        // SAFETY: `u64` is a plain old datatype so we can always transmute from it
```

**Comment by fbstj** [comment]
are the six now-safe intrinsics (`size_of_val`, `min_align_of_val`, `discriminant_value`, `type_id`, `likely`, `unlikely`) blocked on different PRs?

**Comment by oli-obk** [comment]
> are the six now-safe intrinsics (size_of_val, min_align_of_val, discriminant_value, type_id, likely, unlikely) blocked on different PRs?

No, I just did them here because otherwise I'd have had to add `// SAFETY: this intrinsic is safe to call, it's just unsafe because noone made it safe`

**Comment by JohnCSimon** [comment]
Ping from triage: 
@oli-obk @fbstj @KodrAus 
Hi! This has sat idle for over two weeks. What else needs to happen for this pr to move onwards?

Thank you!

**Comment by Centril** [comment]
It's no secret I think this PR is sorely needed.

> Most of the safety comments I'm seeing sort of just trivially reference the line above it so I'm not really sure how a comment actually bolsters understanding of the function because it's all being read anyway.

I don't think each and every safety comment needs to be super helpful and if there are trivially true cases then those can just say "trivially safe". Often however, it's not as trivially true as one experienced reviewer, e.g., @alexcrichton, who often looks at `unsafe` blocks might think. I think a better question is: are these actively harmful?

> which I feel like taken in isolation isn't actually right, in theory that requires some form of alignment checks as well but in reality the function being called here is doing all that internally (it's just the comment that's a bit misleading).

This suggests that it's in fact not as trivial as one might think and that *more elaboration* is needed rather than not having a comment at all.

It's also not the trivial cases that we wish to catch with these comments but it's also not possible to distinguish them without far more complicated semantic analysis (proofs in e.g. Coq). I think the cost of a bit of boilerplate in trivial cases is well worth it if we get to have commentary on more complicated cases. One might say that this problem should be solved with good reviews. To me however, it is clear that reviews have not been satisfactory thus far.

Also, from a language team perspective, I find it really important to know what assumptions the standard library makes on our operational semantics.

**Comment by alexcrichton** [comment]
@rfcbot fcp close

I'm going to propse that we close this. Other members of the libs team can of course disagree! I [explained some reasoning above](https://github.com/rust-lang/rust/pull/63793#issuecomment-524902926).

@Centril I personally think that annoying lints can be actively harmful because they encourage the wrong coding patterns and give a false sense of security or correctness. An `unsafe` block is no more correct if a comment is indicated as to why it's ok, nor is it even any more correct if the comment is quite long and tries to explain everything. Fundamentally `unsafe` blocks will be mistakes eventually and while we can try to head it off I see this as generating too much work for not enough gain. 

No one writing these `unsafe` blocks has the full context of the language in their head and runs through a checklist of "does every single guarantee Rust provides stay upheld?" as they either write `unsafe` or write a comment. You say it's good to know what assumptions the libs team is making, and I can say from my perspective I don't even know what assumptions I'm making when writing `unsafe`. It's not like we're trying to nefariously write code behind the lang team's back that is a gotcha for whatever unsafe guidelines are proposed. In reality we're just trying to get things done and most of this code is years old and probably didn't have a better alternative at the time, but everyone will basically always welcome centralization of `unsafe` to have fewer and fewer `unsafe` blocks.

**Comment by Centril** [comment]
@alexcrichton 

> @Centril I personally think that annoying lints can be actively harmful because they encourage the wrong coding patterns [...]

Do you think documenting why an `unsafe { ... }` block is correct is the wrong coding pattern? I think that's something great and which should be encouraged.

> [...] and give a false sense of security or correctness.

This feels to me like "because we cannot have perfect security or correctness we should not try". I'd like to point to @RalfJung's comments [1](https://internals.rust-lang.org/t/pre-rfc-procmacros-implemented-in-wasm/10860/31) and [2](https://internals.rust-lang.org/t/pre-rfc-procmacros-implemented-in-wasm/10860/31) which feel relevant.

> An `unsafe` block is no more correct if a comment is indicated as to why it's ok, nor is it even any more correct if the comment is quite long and tries to explain everything.

Comments are not intended as a *proof* of correctness. However, they can at least be read by humans during review and after the fact. I believe one can use good judgement in the length one gives to unsafe comments. No one has suggested that one should "explain everything".

> Fundamentally `unsafe` blocks will be mistakes eventually and while we can try to head it off I see this as generating too much work for not enough gain.

I can understand that it might be a non-trivial amount for you in libstd's operating system related code. However, in key parts like libcore and liballoc I think it's well worth the time investment.

> No one writing these `unsafe` blocks has the full context of the language in their head and runs through a checklist of "does every single guarantee Rust provides stay upheld?" as they either write `unsafe` or write a comment.

No one has suggested that you read through the entire nomicon each time you write `unsafe { ... }` but there are often some specific things to think through depending on what parts of the language you are using.

> You say it's good to know what assumptions the libs team is making, [...]. It's not like we're trying to nefariously write code behind the lang team's back that is a gotcha for whatever unsafe guidelines are proposed.

I did not say "what assumptions the *libs team is making*". I think most of the code added to the standard library are in fact added by people outside the libs team. I referred to assumptions of *the standard library* as a whole. This is useful information when doing future language design or just for internal consistency of across the standard library.

> [...], and I can say from my perspective I don't even know what assumptions I'm making when writing unsafe. [...]

Perhaps we interpret "assumptions" differently, but I would hope that you at least consider what invariants you must uphold in an `unsafe { ... }` block. If not, that does not inspire confidence.

> In reality we're just trying to get things done and most of this code is years old and probably didn't have a better alternative at the time, but everyone will basically always welcome centralization of `unsafe` to have fewer and fewer `unsafe` blocks.

*"Getting shit done"-ism* is not an excuse nor in opposition to being deliberate and careful with `unsafe { ... }`. Nor is adding comments in opposition to reducing the amount of `unsafe` blocks through refactoring (who is against that?). I am not suggesting that we do everything over night, but we can improve and audit the years-old code over a longer period of time.

**Comment by sfackler** [comment]
> I think that's something great and which should be encouraged.

This does not *encourage* documentation of the safety of `unsafe` blocks, it *mandates* it.

**Comment by Centril** [comment]
@sfackler 

> This does not _encourage_ documentation of the safety of `unsafe` blocks, it _mandates_ it.

Strictly speaking, tidy has a way to opt-out per file.

Overall it is true however that it mandates documentation, but I think that's a good idea as well. Otherwise, people will oftentimes not do it when it is necessary and I think that's a problem.

**Comment by alexcrichton** [comment]
I don't think you'll find anyone on any Rust team who actively says `unsafe` blocks should not be documented, and it seems somewhat disingenuous to think that's where I'm coming from? I am specifically objecting to what @sfackler mentioned, the mandate and requirement to document all `unsafe` blocks.

The amount of work to document `unsafe` blocks in libcore/liballoc is going to be the same whether or not we have this lint, so if that's the goal of the work why not proceed with that without the lint? I've got what I believe are legitimate concerns about the lint actively hindering and causing problems in some areas, but it's not like I'm trying to say "remove all the documentation from unsafe blocks".

**Comment by Centril** [comment]
@alexcrichton 

> I don't think you'll find anyone on any Rust team who actively says `unsafe` blocks should not be documented, [...]

I hope you are right but I think that's a low bar to set.

> [...], and it seems somewhat disingenuous to think that's where I'm coming from?

Elaboration re. what you mean by "wrong coding patterns" would be helpful then as your argument largely felt like "this is security theater".

> The amount of work to document `unsafe` blocks in libcore/liballoc is going to be the same whether or not we have this lint, so if that's the goal of the work why not proceed with that without the lint? 

Because once we have done so, without lints, the documentation coverage would regress over time. If tidy yells at you, even if we have a local override like `// ignore-tidy-safety` on an `unsafe { ... }` block, at least the developer will be [*nudged*](https://en.wikipedia.org/wiki/Nudge_theory) in the direction of providing the comment. In my view, as `unsafe { ... }` is a dangerous thing to do (I mean... it says so on the tin...), it is well justified to lint in this case.

> I've got what I believe are legitimate concerns about the lint actively hindering and causing problems in some areas, [...]

I think those concerns have not been fleshed out. I replied re. the specifics of alignment and whatnot and would appreciate fewer generalities in return. Let's be specific and concrete about what those areas are where problems and hindrances occur (like... are there specific file paths you'd not like to lint, like in the OS / FFI code?). Perhaps we can find a middle ground.

**Comment by sfackler** [comment]
After looking through the diff here, 90+% of them look like "SAFETY: we performed the check that's one line above this unsafe block", or "SAFETY: it's okay to do this". These read to me like line noise - they add no real additional information to the context. It reminds me of Javadoc requirements in some of my college homework assignments where you end up with hundreds of lines of `@param people The people`.

Every PR is code reviewed by another person. It's their job to make sure that the code is sufficiently comprehensible. Sometimes there are tricky invariants or behavior that require a couple of lines of explanation in comments. Sometimes those happen to be oriented around unsafe blocks, but they often aren't.

If someone comes along and is confused by some piece of code, that may result in someone making a PR adding some extra commentary on it, but again, that doesn't have anything to do with what kind of expression that code is.

**Review by dtolnay — APPROVED** [review]
Thanks for working on this @oli-obk.

I read all the "SAFETY:" comments added in this PR and there aren't any that I would mind having in the codebase. I would be happy with merging this.

<br>

I find this comment from @Centril compelling:

> It's also not the trivial cases that we wish to catch with these comments. [...] I think the cost of a bit of boilerplate in trivial cases is well worth it if we get to have commentary on more complicated cases. One might say that this problem should be solved with good reviews. To me however, it is clear that reviews have not been satisfactory thus far.

<br>

In response to @alexcrichton's and @sfackler's similar concerns,

> Most of the safety comments I'm seeing sort of just trivially reference the line above it so I'm not really sure how a comment actually bolsters understanding of the function because it's all being read anyway. 

> After looking through the diff here, 90+% of them look like "SAFETY: we performed the check that's one line above this unsafe block", or "SAFETY: it's okay to do this". These read to me like line noise - they add no real additional information to the context.

I understand this reaction but it's meaningful to call out when the check on the previous line is the only consideration in an unsafe block being okay, as opposed to some other ambient invariant that comes in from elsewhere. We could reserve comments for only those unsafe blocks with a requirement more complicated than the previous line, but when touching nearby code it's a big help to be told when the safety reasoning is all just local.

**Comment by RalfJung** [comment]
I am fully in favor of this PR! `unsafe` operations are marked that way because they have extra preconditions, and an `unsafe` block is an indication by the programmer to the compiler "yes I checked those preconditions". It only seems reasonable to ask for a brief comment (a) showing that this check really happened, and (b) helping others to follow the argument and audit that the check was done correctly.

I am sure @Shnatsel's safety dance would benefit a lot from such conventions being more widely applied in the ecosystem.

> Most of the safety comments I'm seeing sort of just trivially reference the line above it so I'm not really sure how a comment actually bolsters understanding of the function because it's all being read anyway.

I think it is extremely informative that the author of an unsafe block thought that this unsafe block is justified by just locally checking the line(s) above. That makes review just so much easier: if, after a local check, I am *not* convinced that this unsafe block is correct, I don't need to start a complex analysis of the invariants in this file (because maybe the block actually *is* correct but the argument is more complex). I know immediately that the author of the code made a mistake, saving a lot of time.

The comment saying "correct because we just checked X" takes very little time to write, and documents an important piece of information that makes audits a lot simpler. I think that's a good comment.

> There's also other comments like "SAFETY: transmuting a sequence of u8 to u64 is always fine" which I feel like taken in isolation isn't actually right, in theory that requires some form of alignment checks as well but in reality the function being called here is doing all that internally (it's just the comment that's a bit misleading).

So the correctness argument is more subtle than @oli-obk thought; all the more reason to write it down, I'd say.

> No one writing these unsafe blocks has the full context of the language in their head and runs through a checklist of "does every single guarantee Rust provides stay upheld?" as they either write unsafe or write a comment. You say it's good to know what assumptions the libs team is making, and I can say from my perspective I don't even know what assumptions I'm making when writing unsafe. 

I assume that you have *some* argument in your head for why this unsafe block will not cause UB. Sure, the argument does not have to be a full-blown Coq proof, but my expectation is that people don't just write unsafe and *hope* that it'll work out.

Given that, all we ask is that this -- informal, possibly incomplete -- argument be written down. The argument has already been made by the programmer to themselves; putting it down in writing (a) helps them clarify their own thinking, and (b) helps others by not making them re-do the same argument *again* from scratch.

> It's not like we're trying to nefariously write code behind the lang team's back that is a gotcha for whatever unsafe guidelines are proposed. In reality we're just trying to get things done and most of this code is years old and probably didn't have a better alternative at the time, but everyone will basically always welcome centralization of unsafe to have fewer and fewer unsafe blocks.

Nobody is assuming bad intentions here. But how are we supposed to figure out guidelines if we do not even know the kind of reasoning y'all are using to justify the use of unsafe operations to yourself? The PR does not ask that you quote a UCG document for every piece of reasoning, it asks that you give a brief hint for why *you* think this is or should be correct.

So, beyond the benefits listed above, such comments actually are an enormously useful resource to the UCG as they document what some very experience Rust programmers *think* are good rules for unsafe. If that does not align with what we think, we'd like to know so that we can start a discussion. But if your thinking is never written down, we won't notice the mismatch. Doing language design "in the dark", without real-world input, is really hard; such comments would go a long way to help us ground our efforts and make us aware of common and important unsafe idioms.

As @Centril said, there might also be code where this is less useful, such as code interacting closely with OS/FFI interfaces. Let's not lint on that code and instead find areas where it can be more useful. But by sheer volume, that is a minority of the code in this repo, so we have lots of other code that could benefit a lot from such a lint.

> This does not encourage documentation of the safety of unsafe blocks, it mandates it.

> The amount of work to document unsafe blocks in libcore/liballoc is going to be the same whether or not we have this lint, so if that's the goal of the work why not proceed with that without the lint?

Rust has a history of using tools to guide our process, as experience shows that a policy not checked by a tool is most likely not going to be followed very well. We have a large and diverse set of PR authors and reviewers, and at least I am not aware of a "checklist" that reviewers are supposed to follow. So, basically, any repo-wide effort *not* guarded by a tool is unlikely to actually work. This went far enough that when we had PRs that did some repo-wide language editing (full stops and capitalization and things like that), the general response was "those PRs are not worth it unless we have a tool that checks that these mistakes do not creep back in". I would say the same argument applies here: trying to document unsafe blocks is not worth it unless we have a tool that makes sure no new undocumented unsafe blocks can creep back in.

Personally I feel like some checks tidy is doing (such as line lengths or number of lines or banning `TODO`) are mostly annoying, and are fine to be *encouraged* but not good to be *mandated*. But I appreciate that the only way to actually be consistent about such things is to have a tool check them, so I dutifully break my lines and replace TODO by FIXME (or remove it) every time tidy calls me out.

If we *had* a "reviewer checklist", a good first step might be to add a requirement of documentation to that checklist. But we don't, or rather, our checklist is called `tidy`.

You are saying you are fine to *encourage* comments on unsafe blocks, but how, concretely, do you suggest we actually do that in a way that is effective? The way I view it, `tidy` is how we encourage things in this repo.

> After looking through the diff here, 90+% of them look like "SAFETY: we performed the check that's one line above this unsafe block", or "SAFETY: it's okay to do this". These read to me like line noise - they add no real additional information to the context.

@Centril and me have argued above for why that *does* add information. It severely limits the search space of someone auditing the code. It is also very helpful for the UCG to know that *most* unsafe blocks are "trivial"/"local". In fact, I'd even be in favor of having a special keyword for this case so that we can have statistics on this, but that might be asking too much.

> If 90% of unsafe blocks are trivial and don't need comments, this lint is probably over the top.

Disagreed, for the reasons mentioned above.

> If 90% of the comments are on the module instead of unsafe blocks, then literally this PR probably isn't what we want but maybe something similar to it.

Then the unsafe blocks could still use a marker either referencing a module-level invariant or not.

The key important property of `unsafe` in Rust is that every bit of UB in a Rust program is related to some unsafe block somewhere. Sure, it is not always the unsafe block's *fault* (safe code can break invariants it relied upon), but safe code *alone* can never be at fault either. There's always a connection, through some invariant or so, to some unsafe code. That makes unsafe code a great place to *start* a code review: unsafe blocks are like the seed crystals from where the analysis begins; sometimes the analysis can be completed locally because the argument is simple, sometimes the analysis has to grow into surrounding safe code and encompass most of the module; but it is always correct to *start* from the unsafe blocks and then grow from there.

In that process, the hardest part is figuring out just how much information you need to argue for some unsafe block's correctness. Even a brief comment saying "local checks" vs "module-level invariant" goes a *long* way to declare the scope that needs to be considered here, making unsafe-centric review so much simpler.

@Shantsel (and maybe others from safety-dance), you have done quite a bit of review of other people's unsafe code; your input in this matter would be much appreciated.

(I wrote this as I am catching up in this thread, and now realize many of these points have already been raised by @Centril. So I guess what I am saying is, I am in full agreement, and I think this lint would benefit both the quality of code in libcore, would greatly simplify unsafe-centric code audits, and would provide tons of useful input to the UCG and lang team.)

**Comment by Centril** [comment]
Thanks @RalfJung & @dtolnay; excellent points!

I'll make two further ones which I believe have not been raised prior:

1. The standard library code has special requirements and privileges. As such, it is unlikely to be idiomatic Rust. Nonetheless, people do read the standard library code from time to time and take inspiration from it. This gives us an opportunity to encourage the documentation of `unsafe` in the wider crates.io ecosystem.

2. This may be of use for certification efforts for code where MISRA C would be otherwise be used. -- @jamesmunns may be interested?

**Comment by alexcrichton** [comment]
@rfcbot fcp cancel

Ok then, I will not pursue attempting to block this.

**Comment by jamesmunns** [comment]
So, regarding MISRA-ish checks in Rust, I think there would certainly be some people who would want this for the reasons enumerated by Centril. At least the ability to lint on this condition optionally would likely be useful.

Speaking from personal opinions: Best practices like these, particularly in safety critical domains, work best when the tools are supporting a documented best practice. Tools can be useful, but are rarely useful in isolation, or when not addressing the human component of sources of error. It is also important to document how people are expected to interact with the tool and its' feedback, e.g. from the perspective of the person writing the code, or the person reviewing the code.

Rather, if the desire is to improve the state of documentation for unsafe code in core et. al., I would suggest a bit more holistic approach, such as:

1. The expectation that unsafe code must be documented should be added to contributing.md, or a book, or somewhere that is linkable
2. We set the policy for what should be documented in unsafe code docs, and document it somewhere
3. We set a policy for detection and enforcement:
    * The "warning: unsafe code in this PR" bot links to the policy defined in 1/2, prompting the reviewer (perhaps with a checklist based on 2 above) to check for [good] docs, and verify the comment is correct
    * This linker rejects codes that are missing comments wholesale

In a safety critical project, these expectations would be set in a "software development plan", or similar paperwork, and tools like MISRA checkers only serve to help automate those expectations. There do tend to be teams that "cargo cult" certain practices, and miss the forest for the trees, meaning that only the tool is used, and if it passes, it passes. This is how you get lots of `// SAFETY: verification was done manually` comments. The Rust project does not really have overarching and binding guidance like this paperwork, nor the enforcement and review expected in safety critical.

**Comment by gnzlbg** [comment]
In my experience, most `unsafe` code with bugs has comments - what it doesn't have is good tests that trigger the bugs. As a reviewer, I find it super difficult to assess whether the tests added by some PR touching `unsafe` code actually exercise that unsafe code at all or not.

I find comments around unsafe code explaining what whoever wrote the code thought was going on useful, because they make it easy to spot fundamental flaws in their reasoning, but often there aren't any obvious flaws because the source of the bug might be far away in some code that does not even appear on the diff, isn't on the same file, etc. That code might not be even using unsafe, so without tests, a change to that safe Rust code might break this other unsafe Rust code, and since there are no tests, nobody notices.

**Comment by Shnatsel** [comment]
Just checking in to say that @RalfJung raises excellent points and I'm in full agreement.

Based on my experience in [safety-dance](https://github.com/rust-secure-code/safety-dance), the most time-consuming things are to figure out why the unsafe block is actually there in the first place, and whether it relies on a non-local invariant or not. If there is a comment explaining this, auditing safety of the code takes an order of magnitude less time.

**Comment by JohnCSimon** [comment]
Ping from triage - this has sat idle for the last 7 days.
@oli-obk can you please address the merge conflict? 
cc: @dtolnay @KodrAus @alexcrichton 

Thank you.

**Comment by oli-obk** [comment]
I have rebased and resolved any conflicts. I'm going on vacation now, so I won't touch this PR for two weeks or so.

**Comment by JohnCSimon** [comment]
ping from triage
@oli-obk can you please address the merge conflict once you get back from vacation?
Cc: @KodrAus @dtolnay 

thanks

**Comment by oli-obk** [comment]
This is ready for review again

**Review by dtolnay — APPROVED** [review]

**Comment by dtolnay** [comment]
Thank you @oli-obk, this is great.

@bors r+

**Pull request merged**